twister: rework discard handling

nashif opened the pull request:
Add a filter type to distinguish between testsuite filters that are
maintained in the testcase.yaml file and those provided on the command
line.

The issue was that when something is excluded deliberately on the command
line twister would report an error if that filter would filter out an
integration platform. We do not want that, because the filtering is
being done by the caller knowing that some tests would be not run, even
if those are integration platforms.


- samples: hci_spi: remove dependency on hw
- twister: rework how we discard tests

nashif committed:
twister: rework how we discard tests

Add a filter type to distinguish between testsuite filters that are
maintained in the testcase.yaml file and those provided on the command
line.

The issue was that when something is excluded deliberately on the command
line twister would report an error if that filter would filter out an
integration platform. We do not want that, because the filtering is
being done by the caller knowing that some tests would be not run, even
if those are integration platforms.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: hci_spi: remove dependency on hw

This test only runs on one platform, so make sure we do not exclude the
only platform using some other hardware related filters.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

PerMac reviewed (dismissed):
I don't understand the reasoning/logic of this PR:
I cannot understand why some filters are PLATFORM type and others TESTSUITE in most places. What is the key when selecting what type a skip should be? Can this be explained in the comments where you introduce this types? CLI one is rather clear.
I'm also not sure on the part related to integration_platforms. Your PR allows for mixing options for scope selection, which IMO should be mutually exclusive. Or to be precise, integration_mode mustn't be used with any other scope defining arg as it would undermine the point of integration mode. Can you give a use case when this PR would solve an issue? Maybe I am missing sth?

nashif commented:
@PerMac 
> Can you give a use case when this PR would solve an issue? Maybe I am missing sth?

`twister -e kernel -p native_posix -T tests/kernel/device/ -G`, look at testplan.json...

```
            "name":"tests/kernel/device/kernel.device",
            "arch":"posix",
            "platform":"native_posix",
            "run_id":"80619a8fe9c636b4b02faf2051e3a249",
            "runnable":true,
            "status":"error",
            "reason":"Command line testsuite exclude filter but is one of the integration platforms",
            "log":"",
```

If I am intentionally excluding tests using tags or something else and integration option is enabled, we should not be getting errors like that. -- integration option is not an all or nothing, you should be able to narrow down test scope as much as you want on the command line. If you however are in a test already and you are not being excluded  and the filters in the testcase definition conflict, then sure, this should be an error.

nashif commented:
> I'm also not sure on the part related to integration_platforms. Your PR allows for mixing options for scope selection, which IMO should be mutually exclusive. Or to be precise, integration_mode mustn't be used with any other scope defining arg as it would undermine the point of integration mode.

No, this is not what --integration is about. See above. You should be able to limit scope using the command line and get any errors. Looks like you are trying to make --integration do something it is not supposed to do.

PerMac commented:
I see your reasoning but I am not convinced yet. In your command you use both scope defining args `-G` and `-p native_posix` which are mutually exclusive (you either run on integration scope, or just a single platform`. I believe the logic you are proposing to change works as a safe guard for cases like that, where user made a mistake in the command. Why do you need `-G` in the above call?

nashif commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -3625,98 +3639,97 @@ def apply_filters(self, **kwargs):
                     dep_intersection = ts.depends_on.intersection(set(plat.supported))
                     if dep_intersection != set(ts.depends_on):
-                        discards[instance] = discards.get(instance, "No hardware support")
+                        instance.add_filter("No hardware support", Filters.TESTSUITE)
```
sure, looks like you already got a feel for those types.

nashif commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -3625,98 +3639,97 @@ def apply_filters(self, **kwargs):
 
                 if plat.ram < ts.min_ram:
-                    discards[instance] = discards.get(instance, "Not enough RAM")
+                    instance.add_filter("Not enough RAM", Filters.TESTSUITE)
```
this should be platform as well.

nashif commented:
> I see your reasoning but I am not convinced yet. In your command you use both scope defining args `-G` and `-p native_posix` which are mutually exclusive (you either run on integration scope, or just a single platform. I believe the logic you are proposing to change works as a safe guard for cases like that, where user made a mistake in the command. Why do you need `-G` in the above call?

You can leave the -p native_posix out, I did this just to get results out fast...

Try `twister -e kernel  -G  --save-tests out.json`

PerMac commented:
I know it will produce errors now and just trying to explain I made it intentional this way. The reason was to avoid human-errors like the one in your call. Control over the scope is already not the greatest in twister, hence I added "skip->error" functionality to integration mode to mitigate this issue. With your PR in we will have to be extra cautious in all CI related twister calls, to be sure we are not screwing the scope.  If you are fine with this then I am removing my "change for request"

nashif commented:
> The reason was to avoid human-errors like the one in your call.

where is the human error in my call?

> With your PR in we will have to be extra cautious in all CI related twister calls, to be sure we are not screwing the scope.

How can we screw the scope exactly?

> Control over the scope is already not the greatest in twister,

I am not sure what this means, please elaborate.

PerMac commented:
> where is the human error in my call?

I know your point was about using tags and integration mode, but you also mixed it with --platform. I mean, you just admitted that it was an error, since you wanted to get the command fast. This is exactly the type of human-errors I meant. 

> How can we screw the scope exactly?

e.g. by calling -G together with -p, just like in your example.

> Control over the scope is already not the greatest in twister

If integration mode is not used/no integration platforms provided, then selection rules are getting quite complex and not transparent (e.g. using either "default platform" or platform_allow list). And then the scenarios can get descoped due to dozens of filtration options (e.g. typo in required toolchain won't be easily find. Such configuration will just get skipped and CI will ignore it)

nashif commented:
> I know your point was about using tags and integration mode, but you also mixed it with --platform. I mean, you just admitted that it was an error, since you wanted to get the command fast. This is exactly the type of human-errors I meant.

No, that was not an error. My goal was to show the issue in one single step and without too much noise of other tests. 

> you just admitted that it was an error,

did I really?

> > How can we screw the scope exactly?
> 
> e.g. by calling -G together with -p, just like in your example.

why is this wrong? 

`twister  -p qemu_x86  -T samples/hello_world/ -G` ?

You might want to test all samples and tests that have have qemu_x86 as the integration platform, and if qemu_x86 is not an integration platform, skip those tests...


> If integration mode is not used/no integration platforms provided, then selection rules are getting quite complex and not transparent (e.g. using either "default platform" or platform_allow list). And then the scenarios can get descoped due to dozens of filtration options (e.g. typo in required toolchain won't be easily find. Such configuration will just get skipped and CI will ignore it)

well, without filters we are stuck too, you will have 1000s of failures because you are running code that is not supposed to run. And regarding typos, c'mon, you can also have typos in the code and everywhere for that matter, this sucks, but it is life. I am not sure how a typo in a test or in the code can be attributed to twister itself.

We need to be looking at the results and tracking them overtime and ask the question whenever there is a change or let some test management system should this, right now most of this is not visible directly in CI because nobody really looks at the numbers and the only deciding factor is a "green" check from CI.

My view is that we do not have enough ways to filter and we build/run too much stuff. I acknowledged this just in the last the Testing WG and also emphasized the need for a test management system and an overhaul of the tests themselves. Trying to solve issues in twister and blaming twister for typos is a stretch IMO.

hakehuang commented:
@nashif , I merge this pr with the latest weekly code and I see an issue here. (this issues is also existed in mainline code)
I run below command
```
scripts/twister -p mimxrt1060_evk --device-testing --west-runner=pyocd --west-flash=--board-id=022900000582988000000000000000000000000097969905,--erase --device-serial /dev/ttyACM0 -T tests/lib/cmsis_dsp/transform -o /home/ubuntu/report/rt1060 --inline-logs -N -v -S --platform-reports
```
and the log is
```
ZEPHYR_BASE unset, using "/home/shared/disk/zephyr_project/zephyr_test/zephyr"
Renaming output directory to /home/shared/disk/zephyr_project/zephyr_test/zephyr/twister-out.11
INFO    - Zephyr version: v3.1.0-rc1-59-g72329968026c
INFO    - JOBS: 8
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/shared/disk/zephyr_project/zephyr_test/zephyr/twister-out/testplan.json
INFO    - Using 'zephyr' toolchain.

Device testing on:

| Platform       | ID   | Serial device   |
|----------------|------|-----------------|
| mimxrt1060_evk |      | /dev/ttyACM0    |

INFO    - 21 test scenarios (21 configurations) selected, 1 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    -  2/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.rf64.fpu PASSED (device)
INFO    -  3/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.cf32 PASSED (device)
INFO    -  4/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.cf64.fpu PASSED (device)
INFO    -  5/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.rf16 PASSED (device)
INFO    -  6/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.rf32 PASSED (device)
INFO    -  7/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.rf32.fpu PASSED (device)
INFO    -  8/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.rq31.fpu PASSED (device)
INFO    -  9/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.cf64 PASSED (device)
INFO    - 10/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.cq31.fpu PASSED (device)
INFO    - 11/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.cf32.fpu PASSED (device)
INFO    - 12/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.cq31 PASSED (device)
INFO    - 13/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.rf64 PASSED (device)
INFO    - 14/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.rq31 PASSED (device)
INFO    - 15/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.cf16.fpu PASSED (device)
INFO    - 16/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.cq15 PASSED (device)
INFO    - 17/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.rq15.fpu PASSED (device)
INFO    - 18/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.rq15 PASSED (device)
INFO    - 19/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.cq15.fpu PASSED (device)
INFO    - 20/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.cf16 PASSED (device)
INFO    - 21/21 mimxrt1060_evk            tests/lib/cmsis_dsp/transform/libraries.cmsis_dsp.transform.rf16.fpu PASSED (device)

INFO    - 20 of 21 test configurations passed (100.00%), 0 failed, 1 skipped with 0 warnings in 1649.75 seconds
INFO    - In total 6930 test cases were executed, 0 skipped on 1 out of total 456 platforms (0.22%)
INFO    - 0 test configurations executed on platforms, 20 test configurations were only built.

Hardware distribution summary:

| Board          | ID   |   Counter |
|----------------|------|-----------|
| mimxrt1060_evk |      |        20 |
INFO    - Saving reports...
INFO    - Writing JSON report /home/ubuntu/report/rt1060/twister.json
INFO    - Using 'zephyr' toolchain.
INFO    - Writing xunit report /home/ubuntu/report/rt1060/twister.xml...
INFO    - Writing xunit report /home/ubuntu/report/rt1060/twister_report.xml...
INFO    - Writing target report for mimxrt1060_evk...
INFO    - Run completed

```
from the log I judge that all cases are passing, every thing is fine. however when I check the twister_report.xml file as attaced

there are many timeout issue, which in mainline code base is marked as unknow.
[rt1060.zip](https://github.com/zephyrproject-rtos/zephyr/files/8726144/rt1060.zip)

PerMac commented:
>why is this wrong?
twister -p qemu_x86 -T samples/hello_world/ -G ?

This is wrong because you chose two mutually exclusive options for selecting scope. I don't mean, that it is forbidden because I said so. With `-G` you tell twister to chose integration_platforms and turn skips to errors on those, whenever this list is defined in a testcase.yaml. With `-p` you tell twister to ignore whatever the scope was defined and set it to the platform from the CLI. What is the point of usage `-G` if you want to override it with `-p`?

>You might want to test all samples and tests that have have qemu_x86 as the integration platform, and if qemu_x86 is not an integration platform, skip those tests...

Your command won't work like that (also not with this PR in). It will select all the tests, regardless of integration_platform being there or not, and then try to use qemu_x86 as the scope, skipping everything what won't build on qemu_x86. With `-G` on, you will get errors for platforms other then qemu_x86 that appeared on integration_platforms lists. Without it you will get skips instead. I will just repeat, why using `-G`?
There is currently no option to run only tests that have integration_platfors added.

I am fully aware of all the issues you mentioned, we are really on the same page with it. I am just trying to explain why I believe relaxing the `integration mode` restrictions is a risky idea. I fully agree that we need to refine testing scope and I know we need filters. I am just giving my arguments, why I consider the current state suboptimal: it is too easy to get a test skipped in a uncontrollably manner. Exactly because how you stated it:
> nobody really looks at the numbers and the only deciding factor is a "green" check from CI

If I remember correctly, you defined Integration_platforms like "the minimal scope that has to be evaluated if integration mode is on" and someone explicitly added integration_platforms to a yaml, we must ensure that they won't get skipped. Since numbers doesn't matter, it is even more important to show that something is wrong.
IMO one way is to start a process for adding integration_platforms to all yamls. We done it in sdk-nrf and believe that it brought many benefits. But I won't force this solution. 
 
> And regarding typos, c'mon, you can also have typos in the code and everywhere for that matter, this sucks, but it is life. I am not sure how a typo in a test or in the code can be attributed to twister itself.

I am not blaming twister for typos. I am taking abut making twister more typo-proof. Typos are human-errors, fells in the same category I mentioned earlier. A person makes a typo, e.g when modifying a yaml. Due to that some filter stops to work correctly, e.g. incorrect toolchain/arc/ram etc. And a platform that was once in a scope is no longer there (or even none of the platforms can be selected) Then regressions enters the main, CI is skipping the corresponding test and we only see the error after 0.5y when some developer realize it by a chance. It already happened sever times. It was one of the reasons why I keep on nagging about those `integration_platforms` and how I think it can be used. Such issues won't happen on integration_platforms, since they should be protected from skips during integration mode. 

And you are proposing to relax this enforcement, by allowing a concession for mutually exclusive options when given in CLI (-G and -p are mutually exclusive as I tried to explain). And this couple (-G and -p) might be not the only one we should not use. 

If I understood your use case correctly you'd like to get the scope from integration_platforms, be able to filter them according to a given CLI parameter and don't bother about the skips. If so:
What if we instead of modifying the integration mode -G (i.e. using the integration_platforms scope AND forbidding skips).
- Either split into 2: e.g. --integration-scope and --integration_mode, each responsible for one of the above conditions. I am not a fan for this as it will break the current interface
- Add only --integration-scope and leave -G as it is.  
With any from the above you won't need -G since you don't care about skips and could use --integration-scope to use only integration_platfors.

enjiamai commented:
Hi @smrtos , could you please help to review this first? It will conflict with your fixes.

nashif commented:
> And you are proposing to relax this enforcement, by allowing a concession for mutually exclusive options when given in CLI (-G and -p are mutually exclusive as I tried to explain). And this couple (-G and -p) might be not the only one we should not use.

I think you are over analyzing and interpreting the intention here. First, I do not agree with the concept of mutual exclusivity you are representing here. If option are designed to be mutually exclusive, twister should just abort immediately and not do anything if conflicting options are being used.

Now to the purpose of this, it is very simple and does not really need some additional command line options re scope and such. We should be able to run in integration mode on a limited scope, i.e. if I am not changing any kernel code, I do not want to run any kernel tests, so we exclude them using tags. Doing so is legal, because we are limiting the scope to what actually needs to be tested. This is something we do in CI and have been doing this for a long time.  Without the change in the PR, we will get errors on the excluded tests saying that they have been skipped although we are in integration mode, but I do not want those tests to be part of the plan anyways, so why should we report errors on something we are not testing?

PerMac commented:
I understand the idea of using -G and tag filtering. I support it. I am just pointing that not all CLI flags are equal. IMO --tags are closer to --testsuite-root: they don't depend on any platform feature directly (although platforms can have limitations in their yamls tags-related).  But I will stand my ground, that using -G and --platform is pointless. Like I wrote, it won't work in the way you stated. There is no clear logic right now defining how this should work. If we turn off errors from skipping integration_platforms the -G flag will be simply ignored when put together with --platform, since the scope for each tests will be limited only to what was given with --platform.

PerMac commented:
I looked again at options which get CLI_FILTER and I think the only problematic one is --platform. I agree that all others can be useful to run together with -G. If e.g. you'll add exclusivity at the arg_parser level for -G and --platform I will accept the PR

nashif commented:
> But I will stand my ground, that using -G and --platform is pointless. Like I wrote, it won't work in the way you stated. There is no clear logic right now defining how this should work.

you should read my comment again, I am not saying this is something we want to have in CI or anywhere else, it could be used to validate configurations when writing tests, there is nothing wrong with it and it exactly does what it is asked to do..

chen-png commented on scripts/twister:
```diff
@@ -1152,22 +1152,22 @@ def main():
-        for i, reason in tplan.discards.items():
-            if options.platform and i.platform.name not in options.platform:
+        for i in tplan.instances.values():
+            if i.status == "filtered" and options.platform and i.platform.name not in options.platform:
```
those two conditions couldn't be combined into one, otherwise, it will print all instances as skipped as below:
```
DEBUG   - qemu_x86                  tests/kernel/threads/thread_apis/kernel.threads.apis SKIPPED: Unknown
DEBUG   - qemu_x86                  tests/kernel/threads/thread_apis/kernel.threads.apis.pinonly SKIPPED: In test case platform exclude
DEBUG   - qemu_x86                  tests/kernel/mutex/mutex_api/kernel.mutex          SKIPPED: Unknown
DEBUG   - qemu_x86                  tests/kernel/mutex/mutex_error_case/kernel.mutex_error_case SKIPPED: Unknown
DEBUG   - qemu_x86                  tests/kernel/mutex/sys_mutex/system.mutex          SKIPPED: Unknown
DEBUG   - qemu_x86                  tests/kernel/mutex/sys_mutex/system.mutex.nouser   SKIPPED: Unknown
```

we should check "filtered" status firstly, then check whether it belongs to our expected platforms.

smrtos replied:
Yes, the test cases of the unfiltered instances will also be printed. This is not right.

nashif committed:
twister: rename add_missing_testscases

add_missing_case_status is more appropriate for this function.
We are adding missing status, not cases.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
@smrtos @enjiamai @chen-png please revisit
